packages/buidler-core/src/builtin-tasks/compile.ts

```diff
@@ -390,380 +508,435 @@ export default function () {
-        output: any;
-        solidityFilesCache?: SolidityFilesCache;
-        force: boolean;
-      },
-      { config }
-    ): Promise<{ numberOfContracts: number }> => {
-      const pathToBuildInfo = await saveBuildInfo(
-        config.paths.artifacts,
-        input,
-        output,
-        compilationGroup.getVersion()
-      );
-      let numberOfContracts = 0;
-
-      for (const file of compilationGroup.getResolvedFiles()) {
-        log(`Emitting artifacts for file '${file.globalName}'`);
-        if (!force && !compilationGroup.emitsArtifacts(file)) {
-          continue;
-        }
-
-        const emittedArtifacts = [];
-        for (const [contractName, contractOutput] of Object.entries(
-          output.contracts?.[file.globalName] ?? {}
-        )) {
-          log(`Emitting artifact for contract '${contractName}'`);
-          numberOfContracts += 1;
+  internalTask(TASK_COMPILE_SOLIDITY_EMIT_ARTIFACTS)
+    .addParam("compilationJob", undefined, undefined, types.any)
+    .addParam("input", undefined, undefined, types.any)
+    .addParam("output", undefined, undefined, types.any)
+    .setAction(
+      async (
+        {
+          compilationJob,
+          input,
+          output,
+        }: {
+          compilationJob: CompilationJob;
+          input: SolcInput;
+          output: any;
+        },
+        { config, run }
+      ): Promise<{
+        emittedArtifactsPerFile: EmittedArtifactsPerFile;
+      }> => {
+        const artifacts = new Artifacts(config.paths.artifacts);
+
+        const pathToBuildInfo = await artifacts.saveBuildInfo(
+          input,
+          output,
+          compilationJob.getSolcConfig().version
+        );
 
-          const artifact = getArtifactFromContractOutput(
-            contractName,
-            contractOutput
-          );
+        const emittedArtifactsPerFile: EmittedArtifactsPerFile = [];
+        for (const file of compilationJob.getResolvedFiles()) {
+          log(`Emitting artifacts for file '${file.globalName}'`);
+          if (!compilationJob.emitsArtifacts(file)) {
+            continue;
+          }
 
-          await saveArtifact(
-            config.paths.artifacts,
-            file.globalName,
-            artifact,
-            pathToBuildInfo
-          );
+          const emittedArtifacts = [];
+          for (const [contractName, contractOutput] of Object.entries(
+            output.contracts?.[file.globalName] ?? {}
+          )) {
+            log(`Emitting artifact for contract '${contractName}'`);
+
+            const artifact = await run(
+              TASK_COMPILE_SOLIDITY_GET_ARTIFACT_FROM_COMPILATION_OUTPUT,
+              {
+                contractName,
+                contractOutput,
+              }
+            );
+
+            await artifacts.saveArtifactFiles(
+              file.globalName,
+              artifact,
+              pathToBuildInfo
+            );
+
+            emittedArtifacts.push(artifact.contractName);
```

Comment:
It's the contract name because it's associated to its file so the FQN can be obtained. But maybe we can change it to be the FQN so that it's more independent, and consistent with the idea of representing artifacts by their FQN.